### PR TITLE
Ck tile engine gemm unit tests exapand test coverage

### PR DESCRIPTION
**Pull request by msaffari-amd**
## Proposed changes

Add more test coverage for implemented unit-tests (using tile_engine). 
Covering data types, layouts, traits, tile sizes, etc
Add problem size to verify each kernel for multiple problem sizes
Add documents and some refactoring.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ ] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [ ] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

### CODE REVIEW

**Review by spolifroni-amd — DISMISSED** [review]

File: test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    "tile_k": {"values": [16, 32, 64]},
+    "warp_m": {"values": [1, 2, 4]},
+    "warp_n": {"values": [1, 2, 4]},
+    "warp_k": {"values": [1]},
+    "warp_tile_m": {"values": [16, 32]},
+    "warp_tile_n": {"values": [16, 32]},
+    "warp_tile_k": {"values": [16, 32]}
+  },
+  "trait_config": {
+    "pipeline": {"values": ["compv3"]},
```
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json]: What about mem pipeline, v4 pipeline and others ?

File: test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    "scheduler": {"values": ["intrawave"]},
```
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json]: Mem pipeline supports `interwave` as well.

File: test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    "epilogue": {"values": ["default"]},
```
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json]: We should test cshuffle epilogue as well.

File: test/ck_tile/gemm_tile_engine/configs/tile_size_quick_config.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    "warp_n": {"values": [2]},
+    "warp_k": {"values": [1]},
+    "warp_tile_m": {"values": [16]},
+    "warp_tile_n": {"values": [16]},
+    "warp_tile_k": {"values": [16]}
+  },
+  "trait_config": {
+    "pipeline": {"values": ["compv3"]},
+    "scheduler": {"values": ["intrawave"]},
+    "epilogue": {"values": ["default"]},
```
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/configs/tile_size_quick_config.json]: As above: we need to test here all pipelines and epilogues, but just less tile sizes.

File: test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json [hunk 5]
```diff
@@ -0,0 +1,33 @@
+{
+  "problem": {
+    "description": "Comprehensive tile size coverage testing (16-256, step 16)"
+  },
+  "test_params": {
+    "problem_sizes": [
+      {"m": 512, "n": 512, "k": 256, "split_k": 1},
+      {"m": 1024, "n": 512, "k": 512, "split_k": 1},
+      {"m": 512, "n": 1024, "k": 512, "split_k": 1},
+      {"m": 1024, "n": 1024, "k": 256, "split_k": 1}
```
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json]: There are no configs with splitk != 1

File: test/ck_tile/gemm_tile_engine/configs/large_datatype_config.json [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  "tile_config": {
+    "tile_m": {"values": [64]},
+    "tile_n": {"values": [64]},
+    "tile_k": {"values": [16]},
+    "warp_m": {"values": [2]},
+    "warp_n": {"values": [2]},
+    "warp_k": {"values": [1]},
+    "warp_tile_m": {"values": [16]},
+    "warp_tile_n": {"values": [16]},
+    "warp_tile_k": {"values": [16]}
```
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/configs/large_datatype_config.json]: FP32 should work with tiles like 256x128x32 and MFMA like 32x32x8

File: test/ck_tile/gemm_tile_engine/CMakeLists.txt [hunk 7]
```diff
@@ -196,42 +231,134 @@ endif()
...
+#    These data types require smaller warp tiles due to larger memory footprint
+set(LARGE_DATATYPE_CONFIG "large_datatype_config")
+set(LARGE_DATATYPE_CONFIG_FILE "${CMAKE_CURRENT_SOURCE_DIR}/configs/${LARGE_DATATYPE_CONFIG}.json")
+set(LARGE_DATATYPES "fp32")
+
+if(EXISTS ${LARGE_DATATYPE_CONFIG_FILE})
+    message(STATUS "Processing large datatype config: ${LARGE_DATATYPE_CONFIG} (fp32)")
+    # fp32: only RCR layout approved for testing
+    message(STATUS "  fp32: testing only RCR layout (other layouts not approved)")
+    build_gemm_test_targets("fp32" "rcr" "${LARGE_DATATYPE_CONFIG}")
```
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/CMakeLists.txt]: Do we not support other layouts for fp32 ??

File: test/ck_tile/gemm_tile_engine/README.md [hunk 8]
```diff
@@ -17,11 +17,84 @@ JSON Config → tile_engine Python scripts → Generated Headers → Test Execut
 ```
 
 - **`--list_kernels`**: Get available kernel configurations from JSON
+- **`--gen_individual`**: Generate all kernel headers in parallel during CMake configuration
 - **`--gen_single`**: Generate individual kernel header for each configuration
```
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/README.md:21]: It's not so quite what is the difference - just the parallelizm ?

File: test/ck_tile/gemm_tile_engine/CMakeLists.txt
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/CMakeLists.txt:1]: @illsilin  or @vidyasagar-amd  could one of you take a look at those changes ?

File: test/ck_tile/gemm_tile_engine/configs/traits_coverage_config.json
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/configs/traits_coverage_config.json:1]: I'm not sure whether this one is needed. I think we need to test all pipelines, epilogues and schedulers with different tile sizes and input problem sizes in order to make sure that all the boundary cases work - mainly handling K iteration tail.

File: test/ck_tile/gemm_tile_engine/configs/comprehensive_coverage_config.json [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  "tile_config": {
+    "tile_m": {"values": [16, 32, 48, 64, 80, 96, 112, 128, 144, 160, 176, 192, 208, 224, 240, 256]},
+    "tile_n": {"values": [16, 32, 48, 64, 80, 96, 112, 128, 144, 160, 176, 192, 208, 224, 240, 256]},
+    "tile_k": {"values": [16, 32, 64]},
+    "warp_m": {"values": [1, 2, 4]},
+    "warp_n": {"values": [1, 2, 4]},
+    "warp_k": {"values": [1]},
+    "warp_tile_m": {"values": [16, 32]},
+    "warp_tile_n": {"values": [16, 32]},
+    "warp_tile_k": {"values": [16, 32]}
```
Comment (aosewski) [inline test/ck_tile/gemm_tile_engine/configs/comprehensive_coverage_config.json]: Don't we have instructions with warp_tile K = 8, 64 and 128 as well ? ( not all available for all data types)

**Review by aosewski — CHANGES_REQUESTED** [review]

File: test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json (hunk 1 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json]: To include all combinations of trait configurations in Comprehensive_config, the total number of kernels would grow significantly, multiplying the number of kernels by the number of trait combinations. For example, if we cover all variations for pipeline, scheduler, and epilogue traits, we’d end up with around 12 combinations, resulting in over 20,000 kernels.
To manage this, I created a separate JSON file called traits_coverage_config, which uses fixed tile sizes to ensure trait combinations are covered efficiently. For the **tile_size**_comprehensive_config, my idea was to focus on covering all tile size combinations using a fixed trait configuration.
It’s possible to make the comprehensive config cover everything, but would result in an extremely large and time-consuming build and test process.

File: test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json (hunk 2 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json]: same comment as:
https://github.com/ROCm/composable_kernel/pull/3025#discussion_r2450959235

File: test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json (hunk 3 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json]: same comment as:
https://github.com/ROCm/composable_kernel/pull/3025#discussion_r2450959235

File: test/ck_tile/gemm_tile_engine/configs/traits_coverage_config.json
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/traits_coverage_config.json:1]: I see, just should consider the number of kernels. I change them.

File: test/ck_tile/gemm_tile_engine/CMakeLists.txt (hunk 7 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/CMakeLists.txt]: crr is not compiled.
ccr and rrr are compiled but not verified.
This appeared after a recent merge with develop. Prior to that, all three configurations had been approved and were working as expected.

File: test/ck_tile/gemm_tile_engine/configs/tile_size_quick_config.json (hunk 4 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/tile_size_quick_config.json]: I update them

File: test/ck_tile/gemm_tile_engine/CMakeLists.txt (hunk 7 above)
Comment (aosewski) [reply test/ck_tile/gemm_tile_engine/CMakeLists.txt]: Please report bug for that. It's not expected.

File: test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json (hunk 1 above)
Comment (aosewski) [reply test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json]: I think then we should have each such comprehensive "suite" for each _working_ pipeline & scheduler & epilogue configuration. Those tests could be run ie daily in our CI - take a look at convolution's large (K?) tests in jenkins file.

File: test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json (hunk 5 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json]: I added them.

File: test/ck_tile/gemm_tile_engine/README.md (hunk 8 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/README.md:21]: Yes, And after I merge again develop, seems there is a bug in new version for --gen-indivisual in gemm_instance_builder, so I switched using --gen_single again as the way used in tile-engine benchmark

File: test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json (hunk 1 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/tile_size_comprehensive_config.json]: I see, I check them.

File: test/ck_tile/gemm_tile_engine/configs/tile_size_quick_config.json (hunk 4 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/tile_size_quick_config.json]: I’ve configured all three setups, and now a total of 384 kernels are generated for this quick configuration, which seems reasonable.
However, some kernels are failing during compilation. For example:
test_gemm_tile_engine_fp16_rcr_tile_size_quick_config_compv4_cshuffle_intrawave_False_False_False_False_256x32x64_2x2x1_16x16x16
This one passes validation during instance creation, but fails to compile due to LDS memory limitations.
I’m not sure what the recommended approach is for dropping such kernels. Do you have any suggestions or best practices for handling these cases?

File: test/ck_tile/gemm_tile_engine/configs/traits_coverage_config.json
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/traits_coverage_config.json:1]: As mentioned in other comment, main issue is that some instances are generated that have compile error and should be managed in a good way to be dropped and continue other ones.

File: test/ck_tile/gemm_tile_engine/configs/tile_size_quick_config.json (hunk 4 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/tile_size_quick_config.json]: It seems that using -k 0 with ninja (e.g., ninja -k 0 ...) allows the build to continue even if some steps fail.

File: test/ck_tile/gemm_tile_engine/configs/traits_coverage_config.json
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/traits_coverage_config.json:1]: I removed this one and are tested in quick_coverage_config

File: test/ck_tile/gemm_tile_engine/configs/tile_size_quick_config.json (hunk 4 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/tile_size_quick_config.json]: add trait coverage to quick test. now they are around 500 kernel for this test

File: test/ck_tile/gemm_tile_engine/configs/comprehensive_coverage_config.json (hunk 9 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/comprehensive_coverage_config.json]: added

File: test/ck_tile/gemm_tile_engine/configs/large_datatype_config.json (hunk 6 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/configs/large_datatype_config.json]: changed to 256x128x32. 32x32x8 is not compiled

File: test/ck_tile/gemm_tile_engine/CMakeLists.txt (hunk 7 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/CMakeLists.txt]: After merging develop in the branch, are verified now !

Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/CMakeLists.txt]: on gfx90a. all layout are compiled. but on gfx950 compile error for non "rcr" layouts

**Review by aosewski — DISMISSED** [review]

**Comment by illsilin** [comment]
So it looks like these changes will add ~30 minutes of test time on gfx90a;gfx942;gfx950. Is that correct?

File: test/ck_tile/gemm_tile_engine/CMakeLists.txt
Comment (illsilin) [reply test/ck_tile/gemm_tile_engine/CMakeLists.txt:1]: Looks fine to me.

**Comment by msaffari-amd** [comment]
> So it looks like these changes will add ~30 minutes of test time on gfx90a;gfx942;gfx950. Is that correct?

I think so, it includes around 260 new tests (making instances, compiling and verifying)

File: test/ck_tile/gemm_tile_engine/CMakeLists.txt (hunk 7 above)
Comment (msaffari-amd) [reply test/ck_tile/gemm_tile_engine/CMakeLists.txt]: fp32 is removed from tile engine for now

**Comment by msaffari-amd** [comment]
> So it looks like these changes will add ~30 minutes of test time on gfx90a;gfx942;gfx950. Is that correct?

We've applied the update, and the number of unit tests is now reduced by half. This significantly speeds up both the compilation and verification processes.

**Review by aosewski — APPROVED** [review]
